src/lib/README.ctable.md

```diff
@@ -0,0 +1,198 @@
+### `ctable` (lib.ctable)
+
+A ctable is a hash table whose keys and values are instances of FFI
+data types.  In Lua parlance, an FFI value is a "cdata" value, hence the
+name "ctable".
+
+A ctable is parameterized for the specific types for its keys and
+values.  This allows for the table to be stored in an efficient manner.
+Adding an entry to a ctable will copy the value into the table.
+Logically, the table "owns" the value.  Lookup can either return a
+pointer to the value in the table, or copy the value into a
+user-supplied buffer, depending on what is most convenient for the user.
+
+As an implementation detail, the table is stored as an open-addressed
+robin-hood hash table with linear probing.  This means that to look up a
+key in the table, we take its hash value (using a user-supplied hash
+function), map that hash value to an index into the table by scaling the
+hash to the table size, and then scan forward in the table until we find
+an entry whose hash value is greater than or equal to the hash in
+question.  Each entry stores its hash value, and empty entries have a
+hash of `0xFFFFFFFF`.  If the entry's hash matches and the entry's key
+is equal to the one we are looking for, then we have our match.  If the
+entry's hash is greater than our hash, then we have a failure.  Hash
+collisions are possible as well of course; in that case we continue
+scanning forward.
+
+The distance travelled while scanning for the matching hash is known as
+the /displacement/.  The table measures its maximum displacement, for a
+number of purposes, but you might be interested to know that a maximum
+displacement for a table with 2 million entries and a 40% load factor is
+around 8 or 9.  Smaller tables will have smaller maximum displacements.
+
+The ctable has two lookup interfaces.  One will perform the lookup as
+described above, scanning through the hash table in place.  The other
+will fetch all entries within the maximum displacement into a buffer,
+then do a branchless binary search over that buffer.  This second
+streaming lookup can also fetch entries for multiple keys in one go.
+This can amortize the cost of a round-trip to RAM, in the case where you
+expect to miss cache for every lookup.
+
+To create a ctable, first create a parameters table specifying the key
+and value types, along with any other options.  Then call `ctable.new`
+on those parameters.  For example:
+
+```lua
+local ctable = require('lib.ctable')
+local ffi = require('ffi')
+local params = {
+   key_type = ffi.typeof('uint32_t'),
+   value_type = ffi.typeof('int32_t[6]'),
+   hash_fn = ctable.hash_i32,
+   max_occupancy_rate = 0.4,
+   initial_size = math.ceil(occupancy / 0.4)
+}
+local ctab = ctable.new(params)
+```
+
+— Function **ctable.new** *params*
```

Comment:
> Prepend method definitions with the class name. E.g. ctable:new instead of :new

> Why is ctable:new a function? Shouldn't it be a method?

This is a function exported by the module.  It is not a method on an object.  I believe this documentation is correct;  LMK if there are more concerns.